### ISSUE TITLE
Data Validity

### ISSUE THREAD

**Issue by PromInc** [PromInc/organic-search-analytics#10]
The issue has been addressed to me on multiple occasions that the data retrieved by this tool is not accurate when compared to Google Search Console and/or the Google API Explorer.

I am starting this issue in hopes that a community discussion can come out of it to either prove or disprove the theory.

I do see some anomalies in the data, but I'm guessing it's an issue potentially in the Google API.  At the time of writing this the [Google API](https://github.com/google/google-api-php-client) that is being utilized is still technically in beta.

A discussion on this topic that triggered me to start this issue can be found in the comments on [my blog post](http://promincproductions.com/blog/google-api-access-google-search-analytics-from-google-search-console/#comment-22) about this tool.

I ran a few tests on the largest site that I have access to and came up with the following data:

| Data Capture Method | Number of Impressions | Description |
| --- | --- | --- |
| API Explorer | 8,335 | Google API Explorer without a row limit set |
| API Explorer w Row Limit 5000 | 16,250 | Google API Explorer with a row limit set |
| Organic Search Analytics | 16,078 | Utilizing the reporting tool in the Organic Search Analytics |
| Organic Search Analytics DB | 16,078 | Running a query directly in the database from the same data the tool accesses |
| Google Search Console reported | 64,729 | The number listed at the top of a report above the cart |
| GSC download | 11,699 | The sum of impressions found after downloading a report |

![osa-data-validity-gsc](https://cloud.githubusercontent.com/assets/7319505/11129925/c2f8166c-8947-11e5-8e74-41df126ebafd.JPG)
![osa-data-validity-osa](https://cloud.githubusercontent.com/assets/7319505/11129926/c2fdb4b4-8947-11e5-9c2b-f2692af79c4d.JPG)
![osa-data-validity-gapiexp](https://cloud.githubusercontent.com/assets/7319505/11130169/4829d054-8949-11e5-92b7-3bb05cf21a6f.JPG)

One item to note is where in the code you can find what data is passed to the Google API request.
In `organic-search-analytics/inc/code/dataCapture.php`:
[Lines 21-25](https://github.com/PromInc/organic-search-analytics/blob/master/organic-search-analytics/inc/code/dataCapture.php#L21-L25) are the default settings that get used at [Line 139](https://github.com/PromInc/organic-search-analytics/blob/master/organic-search-analytics/inc/code/dataCapture.php#L139).  On this line, we merge in an array of additional override parameters that _can_ be set for advanced users by creating a custom call to `data-capture-run.php`.  Overrides are not applied when using the GUI of the Organic Search Analytics tool.

One last item to point out is acutally quoting something I said in the comments on my blog the other day about how I understand these tools to operate which should explain the differences we find between GSC and OSA/API Explorer in my test above.  But it doesn't really explain why there is a slight difference between the API Explorer and OSA (Google API) that is displayed in my data sample above.

> First off, yes it is common that the data will be slightly different based on how Google has designed their system. Google Search Console provides up to 1,000 records per report. But this tool is utilizing the API access data, which provides 5,000 records per report. That is one of the reasons it pulls data on a daily basis – to ensure that it captures as much data as possible. That’s my reasoning for saying that typically this tool will show more data than GSC would.
> 
> Also to note, when you pull data for one day it actually pulls three reports – Web, Image, and Video. In GSC you can only view one of these at a time, where as in this tool you can view them as combined or separate.
> 
> But probably the biggest reason for the discrepancy is that Google doesn’t even display their data accurately in their own tool – and this is intentional. I heard Google’s own John Mueller explain this scenario within the past month or two but can’t recall exactly where to link to it. My assumption is that your 30k impressions number you found at the top of the GSC page just above the chart. As I understand it, that is a “fairly accurate” number from Google’s eyes and it’s intended to be all encompassing of the real world scenario of what happened on their site. However, if you were to click the download button at the bottom of a GSC report page and total up the impressions that you get in that report (which is the same as what is shown on the website) it will never match up to the number you saw at the top of the page. Again, this is an intentional “feature” that Google provides. Bottom line – they don’t provide you with all of the data.
> 
> For that reason, each and every marketer gets to make their own opinion about if they choose to use this data or not, and if so at what level do we rely on this data. I see it as this data is great at identifying overall site trends. I take the assumption that Google provides a completely random sample of data as opposed to selectively excluding data on an algorithm of what they want to provide you. Also to note, Google Analytics does something kind of similar… They record and provide you all of your data until a report gets too large and then they randomly sample it in your reports. Despite this being frustrating and potentially invalid data, a large majority of businesses rely on Google Analytics to make business decisions.
> 
> I hope that makes sense and helps you to understand the tool and it’s data a bit more – it’s designed to do the best it can with what information it is provided.
> 
> One last thing to point out is as of right now the reporting tool does not display Bing data (if you captured it). It will be added at some point, but for now it collects and stores Bing data only and is holds it until it gets added to the reporting tool.

Please leave your comments below about your personal findings and if you have any suggestions on how this tool could better capture the data from Google to be more accurate to what the API Explorer is providing.

**Comment by Ninjataktikz** [comment]
Hi Prominc,

Any update on this as of lately.

-Gary

**Comment by PromInc** [comment]
I haven't had any time to look further into this.

I was hoping to get more case studies built out on this page to either validate or disprove the discrepancy.  That was a big reason for me showing my work on this - others could recreate the same tests.  Part of it could be that each site could be different given the amount of data, etc. available from Google.

If we can build enough of a case here it'll be easier to open an issue with the [Google PHP API](https://github.com/google/google-api-php-client) and see if that team could find any issues on their end.  It's still in beta at this time and thus it could be an issue there.  But without any feedback here it's hard to build much of a case to tell Google they have an issue.

**Comment by Ninjataktikz** [comment]
Hi Proinc,

Here some more data that I pulled using the new updated version. It is still not accurate to what google search engine pulls.

Both using same query and just showing the top 3. I see huge difference in Impressions from the tool and from Google Search Console.

Using Tool:
![screen shot 2015-12-14 at 12 51 02 pm](https://cloud.githubusercontent.com/assets/8031147/11793683/dd377de6-a261-11e5-8374-a6f57fade3b9.png)

Using Google Search Console:
![screen shot 2015-12-14 at 12 51 29 pm](https://cloud.githubusercontent.com/assets/8031147/11793700/efabe796-a261-11e5-84ee-ff5fa6cd528a.png)

P.S Sorry I had to block out Queries, company policy.

Thanks.

**Comment by PromInc** [comment]
No worries about blocking the queries.  Is it safe that Query 1 from the tool matches to Query 1 in GSC, etc.?

Can you please provide data for the same query from the [Google API Explorer](https://developers.google.com/apis-explorer/#search/webmasters.searchanalytics.query/m/webmasters/v3/webmasters.searchanalytics.query)?

**Comment by PromInc** [comment]
Also as a note, there were no updates to how the data is captured in release 2.3.0.

**Comment by Ninjataktikz** [comment]
Yes Query 1 matches exactly as Query 1 from GSC. I don't have time today but i'll upload Google API Explorer data tomorrow or later this week. Really busy at work atm.

Thanks

**Comment by ctseo** [comment]
@PromInc can you add an calculation on bottom of table with total clicks, total impresions, and avg positions&ctr ?

I have same problem. I see 2 dates with different number of impressions. On GSC i have 416 impresions for date 1-12.12.2015, same to Google APIs Explorer .... and on your tool i have 474 impressions... i think is not huge but for more queries and date timepicker will be a huge difference. On my analitcs report with your tools i see the secound and last 2 dates have different number of impressions. I've checked another query and i see only last 2 dates have difference on impressions.Can you check if can be a bug there ?

**Comment by PromInc** [comment]
@ctseo - can you please try to run the same dates through the [Google API Explorer](https://developers.google.com/apis-explorer/#search/webmasters.searchanalytics.query/m/webmasters/v3/webmasters.searchanalytics.query) and let me know how the data compares?  If you look at the screenshot from my original post in this thread you can see what parameters should be set to mimic what the tool is doing.

To be honest, I'm not surprised when I hear that data doesn't match up 100% to GSC.  As I describe in [this wiki article](https://github.com/PromInc/organic-search-analytics/wiki/Data-Captured-from-Google-Search-Console#data-discrepancies) there is some discrepancy to be expected based on how Google has built their tools.  The API does provide different data than GSC and the totals in GSC is a misleading dataset compared to the rows provided in the same report.  This is by Google's design.

I'm not saying that my code is 100% bug free, but in my trying to troubleshoot this previously I can't find any different way to pull the data that results in the expected data others are asking for.  That's why I'm asking for others to provide solid examples of discrepancies.  I am willing to take this issue to either the Google API team or John Mueller and see if they can look into it unless someone else finds a flaw in the code.  But comparing this tool to GSC really isn't a valid comparison as the way Google built their tools results in different data.

**Comment by ctseo** [comment]
@PromInc : I have checked again now, for another queries and another dates.

Seems to be a bug from exported dates:
-When i set date 1.12 - 12.12 ... last 2 dates seems to be different. 
- If i set date 1.12 - 13.12 - 11&12 december it's different.

I have set to past... in november : all dates it's correct. So... from my opinion works perfect !

You should add on top or bottom of table the total impressions,clicks, avg position and avg CTR.

Next update pls add this... and the posibility to search by url to see the queries ( even if you want exclude some queries).

Here some steps : 
1. Add URL, and show queries, clicks, impressions by specific dates
2. Add URL, exclude some queries ( that start/contains specific words - regex -)

You sould add a filter by how many results to display per page... becouse the browser can crash.
I found a little bug: if you have a long query. Please see the print screen.

![table-error](https://cloud.githubusercontent.com/assets/13134199/11835511/cb946f96-a3dc-11e5-9f52-c9f04787b297.jpg)

**Comment by PromInc** [comment]
@ctseo thank you for your followup and suggestions.

Let me ask you a question regarding the screenshot bug you provided.  I think I know the answer on how to best handle this, but I'm curious to hear your (or others) feedback.  The reason this happens is there technically are 3 tables side by side that create the data grid.  The reason for this pre-dates this tool - it was laid out that way for a different tool I built and copied it from.  I needed to copy the clicks, impressions, etc. but didn't need the primary row.  So the multiple tables allowed me to select/copy only those rows.  Because of this design, the height doesn't flow across the table like it should.  All that said, my question to you is should this be one table instead of broken up tables?  Do you use the table to copy data out of?  I do envision an export to CSV feature on this page at some point which would make this long paragraph a moot point really...  Thoughts?

**Comment by ctseo** [comment]
@PromInc : Yes, this bug it's because you have 3 tables side by side, instead one and better :).
You could use only 1 table width:100% with many rows and columns ( if you want to can add boostrap ).

First row will be the title of columns ( Dates,Queries, Clicks,Impressions,CTR,Position)
Second row can be the total clicks, queries, impressions, avg ctr, avg position )
And next rows will have values.

You should fix this... and after we can make css again for this tool :) Simply change the code .. and for css we will have solutions !

**Comment by PromInc** [comment]
Google recently announced that an update occurred to how they log data that may increase clicks and impressions.
[Google Documentation](https://support.google.com/webmasters/answer/6211453#search_analytics)
[Search Engine Land Announcement of Change](http://searchengineland.com/google-changes-how-they-log-search-analytics-data-as-of-december-14-2015-238958)

@Ninjataktikz I'm interested to know if this changes the anomalies you are seeing.  It'd be great to see data compared from **Organic Search Analytics**, Google Search Console, and the Google API Explorer for the same date/queries.

**Comment by ctseo** [comment]
Hello @PromInc ,

I've checked now, and i still have this differences on old data. I've checked after 14/12/2015 ... and seems to be ok. Will see later...

**Comment by ctseo** [comment]
Well... seems to not be same as GSC or Google ApiExplorer.
GSC info:
Total clicks
12,289
Total impressions
198,745
GoogleApiExplorer:
{
 "rows": [
  {
   "clicks": 12289,
   "impressions": 198745
  }
 ]
}
I think it's to big the differences... !

**Comment by PromInc** [comment]
These numbers match up... what did Organic Search Analytics report?

**Comment by ctseo** [comment]
Sry, i forgot to add Organic Search Analytics stats.  

Clicks: 4066
Impressions: 94280

Do you know why it's big differences ?

**Comment by PromInc** [comment]
I did find a bug in the MySQL import.  I apparently omitted the pesky `addslashes` call, and thus queries with special characters where thus erroring out silently.

Please download the update, 2.3.1 from the master branch and test it.  Let me know if you see any issues before I commit this to a new release.

In troubleshooting this issue, I did verify that the data **Organic Search Analytics** pulls from the **Google API** matches 100% to the same query in the **Google API Explorer**.  In other words, we are capturing data correctly from Google to the tool.  But the breakdown here was getting that data into the database.

Please let me know if you are seeing more accurate data on your end.  This of course will only work for future data captures from Google or Bing.  I'll try to post the steps that you could take to try and re-capture some of your past data if you choose - that's a "proceed at your own risk" sort of thing.

**Comment by ctseo** [comment]
I've changed the code from wmtimport.php with the new one and i've deleted the dates between 10/07-01/01 and nothing changed. I see same results as last version.

My sql command to delete rows : FROM `search_analytics` WHERE DATE(`date`) BETWEEN DATE('2015-10-07') AND DATE('2016-01-01')

I have did something wrong ? Only the wmtimport.php file was updates... right?

**Comment by PromInc** [comment]
Correct, only `wmtimport.php` has changed.
- [wmtimport.php code change for Google](https://github.com/PromInc/organic-search-analytics/commit/1a519385df5b117fbface8a391efbac314d59342)
- [wmtimport.php code change for Bing](https://github.com/PromInc/organic-search-analytics/commit/31e3bcd34aa7c6472ce6d462086d89ba96f4cae4)

##### Debugging

Lets try something different here then.  I've created a debug mode for the data import process on a new branch: [data-validity-debug](https://github.com/PromInc/organic-search-analytics/tree/data-validity-debug)

The only file in this branch that has updated is `wmtimport.php`.

To use this change, run the following URL in your web browser:

```
http://localhost/organic-search-analytics/data-capture-run.php?type=goog
leSearchAnalytics&domain=yourwebsite.com&date=2015-12-30
```
- Change the path to your installation directory
- Change the domain name to that which you are testing (this must match exactly to what is in Google Search Console)
- [More information on the parameter options](https://github.com/PromInc/organic-search-analytics/wiki/Data-Capture-Parameters)

> WARNING - this will import data for the date selected and thus will have skew your data for that day.  This should be performed on a test database/installation else be prepared to delete the false data that is captured for that day.

When you run this page, you will see output from the data that is being processed into MySQL.  It's a big loop, so here are the lines of information you'll see:
- Three search engine types are being imported from Google - web, image, video.  The output repeats for each type and is indicated by a heading that says `Import data for search type: {Search Type}`
- The record number for that search type is displayed - the highest you'll get to is 4999 as a maximum of 5000 rows is allowed per request from Google's API
- The query that MySQL is attempting to perform is displayed.
- Whether the query performed successfully or not - it'll display either `bool(true)` or `bool(false)`
- If the above line displays `bool(false)`, then the details of the error from the MySQL request should display.

Look through this page and see if your find any `bool(false)` lines and then the corresponding errors that follow.

Also, at the bottom of each search type section it'll display the total number of records imported for that search type.  That number should be 1 more than the last record# line.  (record#'s are a zero based index, the number imported are a 1 based index, hence why they are not exactly the same).

This of course is looking at data one day at a time (I know your comment was from a larger date range) but this will help us to isolate where the problem is.

Let me know what you find for errors here.

Thanks.

**Comment by ctseo** [comment]
I dont know if i did something wrong, but ... i have only bool(true)  !

**Comment by PromInc** [comment]
Are you saying it only shows once?  `bool(true)` means that the data was imported correctly.  This is a good sign, but it should show multiple times - once for each record that is imported.

This process was to try and validate that the data imported is accurate because you stated:

> I've changed the code from wmtimport.php with the new one and i've deleted the dates between 10/07-01/01 and nothing changed. I see same results as last version.

To really clear the air, I've written a wiki post that explains how I'm comparing data between different reporting tools.
[Validating Data Provided by Google](https://github.com/PromInc/organic-search-analytics/wiki/Validating-Data-provided-by-Google)

I suggest you follow the two methods outlined in the **[Standard Practice](https://github.com/PromInc/organic-search-analytics/wiki/Validating-Data-provided-by-Google#standard-practice)** section.  This of course assumes that you have deleted your data and used the tool to re-pull the data from Google using the Data Capture page AFTER updating to either 2.3.1 or 2.3.2.

To follow these steps however, you'll need to install one more update that I just merged into the Master Branch - version 2.3.2.  I think you'll like the updates to the report page from this update.

Let me know if this doesn't make sense.  There's a lot of information both here and on the Wiki page, but this is how I've been validating my data while testing this.

**Comment by PromInc** [comment]
I've added a deletion page that will allow you to mass-delete data from past dates since much of it is inaccurate (depending on your search queries) given the bug found and fixed yesterday.  The delete option is available in version 2.3.4 which is the current master branch.

There is a link to this page from the home page.  I didn't want to make it too accessible because I don't advise the standard user using this all that often.  This would be a good time to run it so as to ensure as much data as is available from Google is up to date.  Also if you feel you have a bad data pull from Google for one reason or another you could use this feature to clear out a day or two and re-pull the data.

If you want to try and update your data do the following:
1. Ensure you have the most up-to-date version (2.3.4) of **Organic Search Analytics** by downloading the master branch and installing it into your repository.  Remember to not delete your _config_ folder however.
2. From the home page, choose **Delete Data**
3. Select the dates and domains you wish to delete the data for
4. At the bottom of the page choose **Yes** and then **Delete Records**
5. Once the process is done, go to **Data Capture** from the main menu
6. Click **Run** to capture all domains data for all dates, or choose specific days/domains you wish to capture.

**Comment by ctseo** [comment]
Hello @PromInc  , i'm back ! 

I dont know if i have a problem mysql or php version. I've updated your soft ( downloaded new files and replaced from master branch .

Here's a print-screen with results of GSC, GApiExplorer, Tool v2.3.6, Tool v2.3.1 !

![print-toolorganic](https://cloud.githubusercontent.com/assets/13134199/12230171/0e9f1d02-b857-11e5-957b-fbdb604873b7.jpg)

Seems the bug did not gone.

P.S: Save this Report to Quick Links button doesnt work :) !

**Comment by PromInc** [comment]
Hi @ctseo, thank you for providing the screenshots - very helpful in looking this over.

A few things:
- Under Google Search Console and Google API Explorer you have chosen **web** as the search type - however in the **Organic Search Analytics** screenshots you have chosen **All** (web, image, video).  For that reason the comparison is not the same and thus the data shouldn't match.
- As discussed in the Wiki article regarding [Validating Data provided by Google](https://github.com/PromInc/organic-search-analytics/wiki/Validating-Data-provided-by-Google#google-search-console-webmaster-tools), the numbers from Google Search Console really don't mean much in this comparison.  Yes, they match what the API Explorer returns, but this is a total overview and can not be recreated for any sites with more than 5,000 queries per day due to how Google has built their API.  This is by design of Google.  Because these total numbers are inaccurate I have chosen to not even capture them in the method you have from the API as they just result in misleading data.  Instead, **Organic Search Analytics** shows the total of all of the data it does have saved, which is the data associated with a particular query per day per search type.
  ![google-search-analytics-data-comparison](https://cloud.githubusercontent.com/assets/7319505/12234996/52c474ee-b836-11e5-80c8-c441046cc536.jpg)
- Continuing off of the previous point, the Google API Explorer example you showed does not match the method in which **Organic Search Analytics** queries the data from the Google API.  I have outlined in detail how the data is captured and how you can break down the response to compare it accurately to **Organic Search Analytics** in the Wiki post [Validating Data provided by Google](https://github.com/PromInc/organic-search-analytics/wiki/Validating-Data-provided-by-Google#google-api-explorer).  I highly encourage you to follow those steps if you are going to try and make this comparison and validation.  If you don't follow those steps you aren't looking at the data available to you in the same aspect and then the comparison will not be accurate.  Your API request was looking for total information - not on a per query per day per search type method.

One thing that does seem odd however is the two screenshots you provided form **Organic Search Analytics**.  

Regarding the provided screenshots from **Organic Search Analytics**, I have a few comments.  I'm looking at this with the assumption that you had deleted and re-captured data between the screenshots from 2.3.1 to 2.3.6.
- Looking at 2016-01-02 and 2016-01-03, this data looks great in my mind.  There are more impressions/clicks in 2.3.6, which implies to me that the bug fix worked correctly.  Again, the bug fix was to capture queries that had special characters (quotes, colons, slashes, etc.) so an uptick is obviously expected, but it shouldn't be drastically larger.
- Looking at 2016-01-01 we see no change.  This could be because you didn't delete data and re-capture for that date (you didn't specify, so an assumption I'm making) or it could be that there were no queries that day with special characters.  Given that the uptick on 1-2 and 1-3 aren't that drastically large, that could be a true statement for your site.
- Looking at 2016-01-04 however, this data set is a bit strange.  The numbers dropped drastically which is not the expected outcome of course, however it was oddly high in 2.3.1 given the other days around it.  Given that the **queries** column matches exactly between the two versions, I believe what happened is that somehow the data for 2016-01-04 was captured 3 times from Google.  Dividing the impressions and clicks from 2016-01-04 by 3 matches the data pattern for what it should be that day.  There was another bug that was fixed in 2.3.5 where the date range being calculated for days that need data from Google was off by 1 day, and thus a day would continue to show up on the data capture screen at times saying it needed data.  If that was happening and you were on that screen often enough, there is a chance you could have clicked it more than once and captured additional data that was already captured.  That would result in what we are seeing here.

So bottom line, the screenshots from **Organic Search Analytics** look like a great sign in my opinion that things are working correctly.  The screenshots from the Google tools don't give me anything accurate to compare against however.  If you want to further validate the data, I highly suggest you follow the detailed instructions outlined in the Wiki post as was previously suggested prior.  [Validating Data provided by Google](https://github.com/PromInc/organic-search-analytics/wiki/Validating-Data-provided-by-Google)

**Comment by ctseo** [comment]
@PromInc : can you make a total of results directly comes from API without import to MYSQL? Try to make a sample text export from API and make a total clicks or numbers ... i think we can have 2 bugs here: 1 when we try to import dates in MYSQL and 2'nd when you try to get all reports from Google API. I dont think ... I do not think googleAPI gives us information wronged their tool ( GSC ). It must be something wrong here... 

![gsc-tool2](https://cloud.githubusercontent.com/assets/13134199/12236406/389449f4-b882-11e5-8f75-01bb688ae83f.jpg)

Now you see all with Web search type and ALL on device.

**Comment by PromInc** [comment]
But all you are pulling is totals per day.  This doesn't provide you with any granularity on a specific query - which is the focus of this tool.  Totals mean nothing if you can't quantify what the sum of that data is.

Google only provides a maximum of 5,000 rows in a given API request.  Your examples result in 1 row.  So yes, they will give you what you requested - a total for the requested information.

What **Organic Search Analytics** does however is asks for totals per query.  If your site has more than 5,000 queries per day per search type then the totals will never match how you are accessing your data.  And even if you have less than 5,000 queries per day per search type I'm not 100% confident that they will return all of the data.

What you are calling bugs I call a difference in how you want to view data.  This tool could be modified to pull totals like you are showing here, but all that is an large number with no way to recreate it in the granular data.

The point of this issue was to identify and correct bugs related to how the tool is currently pulling data - on a per query per day by device basis.  Your "bugs" are not related to this issue as it appears you are looking for a simple total.

**Comment by ctseo** [comment]
I've checker again now ... but only per query, and all data was exactly like in GSC.

So, in conclusion, @PromInc  you are the BEST! :+1: 

I see now version 2.3.7 ! What's new ?

**Comment by PromInc** [comment]
Glad to hear it's working for you @ctseo.

Version 2.3.7 was a minimal update - added the number of dates selected in a date range on the report page.

I'm going to close out this issue - it appears the issue has been resolved.
